File: rust/src/range.rs
pub struct Range {
  pub start: usize,
  pub end: usize,
}

impl Range {
  pub fn new(start: usize, end: usize) -> Self {
    Self { start, end }
  }

  pub fn length(&self) -> usize {
    self.end.saturating_sub(self.start)
  }

  pub fn copy(&self) -> Self {
    Self { start: self.start, end: self.end }
  }

  pub fn to_string(&self) -> String {
    format!(
      "#<Range [{}, {}]>",
      self.start,
      self.end,
    )
  }
}

File: node/lib/range.ts
export class Range {
  readonly start: number
  readonly end: number

  constructor(start: number, end: number) {
    this.start = start
    this.end = end
  }

  toArray(): number[] {
    return [this.start, this.end]
  }

  toJSON(): number[] {
    return this.toArray()
  }

  treeInspect(): string {
    return this.toArray().toString()
  }

  inspect(): string {
    return `#<Range ${this.toArray()}>`
  }

  toString(): string {
    return this.inspect()
  }
}
